File: engine/src/runtime/function/input/input.cpp
// do@Redlive

#include "input.h"

#include "input_manager.h"

namespace dodoe {

    InputManager* Input::input_manager_ = nullptr;

    void Input::initialize(InputManager* input_manager) { input_manager_ = input_manager; }
    void Input::shutdown() { input_manager_ = nullptr; }

    Bool Input::RegisterActionMap(const String& map_name, Int priority) {
        return input_manager_ && input_manager_->registerActionMap(map_name, priority);
    }
    Bool Input::UnregisterActionMap(StringView map_name) {
        return input_manager_ && input_manager_->unregisterActionMap(map_name);
    }
    Bool Input::SetActionMapEnabled(StringView map_name, Bool enabled) {
        return input_manager_ && input_manager_->setActionMapEnabled(map_name, enabled);
    }
    Bool Input::SetActionMapConsume(StringView map_name, Bool consume) {
        return input_manager_ && input_manager_->setActionMapConsume(map_name, consume);
    }
    Bool Input::PushInputContext(StringView map_name) {
        return input_manager_ && input_manager_->pushInputContext(map_name);
    }
    Bool Input::PopInputContext(StringView map_name) {
        return input_manager_ && input_manager_->popInputContext(map_name);
    }
    Bool Input::RegisterAction(StringView map_name, const String& action_name, InputActionValueType value_type) {
        return input_manager_ && input_manager_->registerAction(map_name, action_name, value_type);
    }
    Bool Input::BindKey(StringView map_name, StringView action_name, KeyCode key, Float scale) {
        return input_manager_ && input_manager_->bindKey(map_name, action_name, key, scale);
    }
    Bool Input::BindKey2D(StringView map_name, StringView action_name, KeyCode key, Vector2f scale) {
        return input_manager_ && input_manager_->bindKey2D(map_name, action_name, key, scale);
    }
    Bool Input::BindMouseButton(StringView map_name, StringView action_name, MouseCode button, Float scale) {
        return input_manager_ && input_manager_->bindMouseButton(map_name, action_name, button, scale);
    }
    Bool Input::BindMouseDelta(StringView map_name, StringView action_name, Float scale) {
        return input_manager_ && input_manager_->bindMouseDelta(map_name, action_name, scale);
    }
    Bool Input::BindMouseWheel(StringView map_name, StringView action_name, Float scale) {
        return input_manager_ && input_manager_->bindMouseWheel(map_name, action_name, scale);
    }
    Bool Input::SetBindingInteraction(StringView map_name, StringView action_name,
                                       InputInteraction interaction, Float hold_seconds) {
        return input_manager_ && input_manager_->setBindingInteraction(map_name, action_name, interaction, hold_seconds);
    }
    Bool Input::LoadActionAsset(const String& absolute_path) {
        return input_manager_ && input_manager_->loadActionAsset(absolute_path);
    }

    Bool Input::SetBindingTapParams(StringView map_name, StringView action_name, Size_t binding_index,
                                    Int tap_count, Float tap_window_seconds) {
        return input_manager_ && input_manager_->setBindingTapParams(map_name, action_name, binding_index,
                                                                     tap_count, tap_window_seconds);
    }
    Bool Input::SetBindingRepeatParams(StringView map_name, StringView action_name, Size_t binding_index,
                                       Float repeat_delay, Float repeat_rate) {
        return input_manager_ && input_manager_->setBindingRepeatParams(map_name, action_name, binding_index,
                                                                        repeat_delay, repeat_rate);
    }
    Bool Input::SetBindingProcessor(StringView map_name, StringView action_name, Size_t binding_index,
                                    const InputProcessor& processor) {
        return input_manager_ && input_manager_->setBindingProcessor(map_name, action_name, binding_index, processor);
    }
    Bool Input::BindGamepadButton(StringView map_name, StringView action_name, GamepadButtonCode button,
                                  InputDeviceId device_id, Float scale) {
        return input_manager_ && input_manager_->bindGamepadButton(map_name, action_name, button, device_id, scale);
    }
    Bool Input::BindGamepadAxis(StringView map_name, StringView action_name, GamepadAxisCode axis,
                                InputDeviceId device_id, Float scale) {
        return input_manager_ && input_manager_->bindGamepadAxis(map_name, action_name, axis, device_id, scale);
    }
    Bool Input::BindGamepadStick(StringView map_name, StringView action_name, GamepadAxisCode stick_axis,
                                 InputDeviceId device_id, Float scale) {
        return input_manager_ && input_manager_->bindGamepadStick(map_name, action_name, stick_axis, device_id, scale);
    }
    Bool Input::BindComposite(StringView map_name, StringView action_name,
                              const DynamicArray<InputBinding>& parts, InputDeviceId device_id) {
        return input_manager_ && input_manager_->bindComposite(map_name, action_name, parts, device_id);
    }
    Bool Input::SetBindingOverride(StringView map_name, StringView action_name, Size_t binding_index,
                                   const InputBinding& binding) {
        return input_manager_ && input_manager_->setBindingOverride(map_name, action_name, binding_index, binding);
    }
    Bool Input::ClearBindingOverride(StringView map_name, StringView action_name, Size_t binding_index) {
        return input_manager_ && input_manager_->clearBindingOverride(map_name, action_name, binding_index);
    }
    Bool Input::GetBinding(StringView map_name, StringView action_name, Size_t binding_index, InputBinding& out) {
        return input_manager_ && input_manager_->getBinding(map_name, action_name, binding_index, out);
    }
    Bool Input::BeginRebindSession(StringView map_name, StringView action_name, Size_t binding_index) {
        return input_manager_ && input_manager_->beginRebindSession(map_name, action_name, binding_index);
    }
    void Input::CancelRebindSession() {
        if (input_manager_) input_manager_->cancelRebindSession();
    }
    Bool Input::IsRebindSessionActive() {
        return input_manager_ && input_manager_->isRebindSessionActive();
    }
    Bool Input::LoadConfigOverrides(const FsPath& project_path, const FsPath& user_path) {
        return input_manager_ && input_manager_->loadConfigOverrides(project_path, user_path);
    }
    Bool Input::SaveUserConfigOverrides(const FsPath& user_path) {
        return input_manager_ && input_manager_->saveUserConfigOverrides(user_path);
    }

    InputActionId Input::FindActionId(StringView map_name, StringView action_name) {
        return input_manager_ ? input_manager_->findActionId(map_name, action_name) : kInvalidInputActionId;
    }
    InputActionId Input::FindActionId(StringView qualified_name) {
        return input_manager_ ? input_manager_->findActionId(qualified_name) : kInvalidInputActionId;
    }

    Bool Input::IsActionDown(StringView action_name) {
        return input_manager_ && input_manager_->isActionDown(action_name);
    }
    Bool Input::WasActionPressed(StringView action_name) {
        return input_manager_ && input_manager_->wasActionPressed(action_name);
    }
    Bool Input::WasActionReleased(StringView action_name) {
        return input_manager_ && input_manager_->wasActionReleased(action_name);
    }
    Float Input::GetActionAxis(StringView action_name) {
        return input_manager_ ? input_manager_->getActionAxis(action_name) : 0.0f;
    }
    Vector2f Input::GetActionVector2(StringView action_name) {
        return input_manager_ ? input_manager_->getActionVector2(action_name) : Vector2f(0.0f);
    }
    Vector2f Input::GetMousePosition() {
        return input_manager_ ? input_manager_->getMousePosition() : Vector2f(0.0f);
    }
    Vector2f Input::GetMouseDelta() {
        return input_manager_ ? input_manager_->getMouseDelta() : Vector2f(0.0f);
    }
    Vector2f Input::GetMouseWheel() {
        return input_manager_ ? input_manager_->getMouseWheel() : Vector2f(0.0f);
    }

    Bool Input::IsGamepadConnected(InputDeviceId device_id) {
        return input_manager_ && input_manager_->isGamepadConnected(device_id);
    }
    Bool Input::IsGamepadButtonDown(InputDeviceId device_id, GamepadButtonCode button) {
        return input_manager_ && input_manager_->isGamepadButtonDown(device_id, button);
    }
    Bool Input::IsGamepadButtonPressed(InputDeviceId device_id, GamepadButtonCode button) {
        return input_manager_ && input_manager_->isGamepadButtonPressed(device_id, button);
    }
    Bool Input::IsGamepadButtonReleased(InputDeviceId device_id, GamepadButtonCode button) {
        return input_manager_ && input_manager_->isGamepadButtonReleased(device_id, button);
    }
    Float Input::GetGamepadAxis(InputDeviceId device_id, GamepadAxisCode axis) {
        return input_manager_ ? input_manager_->getGamepadAxis(device_id, axis) : 0.0f;
    }

    Bool Input::IsActionDown(InputActionId action_id) {
        return input_manager_ && input_manager_->isActionDown(action_id);
    }
    Bool Input::WasActionPressed(InputActionId action_id) {
        return input_manager_ && input_manager_->wasActionPressed(action_id);
    }
    Bool Input::WasActionReleased(InputActionId action_id) {
        return input_manager_ && input_manager_->wasActionReleased(action_id);
    }
    Float Input::GetActionAxis(InputActionId action_id) {
        return input_manager_ ? input_manager_->getActionAxis(action_id) : 0.0f;
    }
    Vector2f Input::GetActionVector2(InputActionId action_id) {
        return input_manager_ ? input_manager_->getActionVector2(action_id) : Vector2f(0.0f);
    }

    InputSubscriptionId Input::Subscribe(StringView action_name, InputActionPhase phase,
                                          InputActionListener listener) {
        return input_manager_ ? input_manager_->subscribe(action_name, phase, std::move(listener)) : 0;
    }
    InputSubscriptionId Input::Subscribe(InputActionId action_id, InputActionPhase phase,
                                          InputActionListener listener) {
        return input_manager_ ? input_manager_->subscribe(action_id, phase, std::move(listener)) : 0;
    }
    void Input::Unsubscribe(InputSubscriptionId subscription_id) {
        if (input_manager_) input_manager_->unsubscribe(subscription_id);
    }

} // namespace dodoe

File: engine/src/scriptcore/Source/Input/Input.cs
namespace GreenCake;

using System;

public enum InputActionValueType
{
    Button = 0,
    Axis1D = 1,
    Axis2D = 2,
}

public enum InputActionPhase
{
    Waiting = 0,
    Started = 1,
    Performed = 2,
    Canceled = 3,
}

public enum InputInteraction
{
    Press = 0,
    Hold = 1,
    Tap = 2,
    MultiTap = 3,
    Chord = 4,
    Toggle = 5,
    Repeat = 6,
}

public enum InputProcessorType
{
    DeadZone = 0,
    Normalize = 1,
    Scale = 2,
    Invert = 3,
    Clamp = 4,
}

public enum GamepadButtonCode
{
    A = 0,
    B = 1,
    X = 2,
    Y = 3,
    LB = 4,
    RB = 5,
    Back = 6,
    Start = 7,
    Guide = 8,
    LeftStick = 9,
    RightStick = 10,
    DpadUp = 11,
    DpadRight = 12,
    DpadDown = 13,
    DpadLeft = 14,
}

public enum GamepadAxisCode
{
    LeftX = 0,
    LeftY = 1,
    RightX = 2,
    RightY = 3,
    LeftTrigger = 4,
    RightTrigger = 5,
}

public static class Input
{
    public static bool RegisterActionMap(string mapName, int priority = 0)
    {
        return NativeCalls.Native_InputRegisterActionMap(mapName, priority);
    }

    public static bool UnregisterActionMap(string mapName)
    {
        return NativeCalls.Native_InputUnregisterActionMap(mapName);
    }

    public static bool SetActionMapEnabled(string mapName, bool enabled)
    {
        return NativeCalls.Native_InputSetActionMapEnabled(mapName, enabled);
    }

    public static bool SetActionMapConsume(string mapName, bool consume)
    {
        return NativeCalls.Native_InputSetActionMapConsume(mapName, consume);
    }

    public static bool PushContext(string mapName) => NativeCalls.Native_InputPushContext(mapName);

    public static bool PopContext(string mapName) => NativeCalls.Native_InputPopContext(mapName);

    public static bool RegisterAction(string mapName, string actionName, InputActionValueType valueType)
    {
        return NativeCalls.Native_InputRegisterAction(mapName, actionName, (int)valueType);
    }

    public static bool BindKey(string mapName, string actionName, KeyCode key, float scale = 1.0f)
    {
        return NativeCalls.Native_InputBindKey(mapName, actionName, key, scale);
    }

    public static bool BindKey2D(string mapName, string actionName, KeyCode key, float x, float y)
    {
        return NativeCalls.Native_InputBindKey2D(mapName, actionName, key, x, y);
    }

    public static bool BindMouseButton(string mapName, string actionName, MouseCode button, float scale = 1.0f)
    {
        return NativeCalls.Native_InputBindMouseButton(mapName, actionName, button, scale);
    }

    public static bool BindMouseDelta(string mapName, string actionName, float scale = 1.0f)
    {
        return NativeCalls.Native_InputBindMouseDelta(mapName, actionName, scale);
    }

    public static bool BindMouseWheel(string mapName, string actionName, float scale = 1.0f)
    {
        return NativeCalls.Native_InputBindMouseWheel(mapName, actionName, scale);
    }

    public static bool BindGamepadButton(string mapName, string actionName, GamepadButtonCode button, uint deviceId = 0, float scale = 1.0f)
    {
        return NativeCalls.Native_InputBindGamepadButton(mapName, actionName, (int)button, deviceId, scale);
    }

    public static bool BindGamepadAxis(string mapName, string actionName, GamepadAxisCode axis, uint deviceId = 0, float scale = 1.0f)
    {
        return NativeCalls.Native_InputBindGamepadAxis(mapName, actionName, (int)axis, deviceId, scale);
    }

    public static bool BindGamepadStick(string mapName, string actionName, GamepadAxisCode stickAxis, uint deviceId = 0, float scale = 1.0f)
    {
        return NativeCalls.Native_InputBindGamepadStick(mapName, actionName, (int)stickAxis, deviceId, scale);
    }

    public static bool BindComposite(string mapName, string actionName, string partsJson, uint deviceId = 0)
    {
        return NativeCalls.Native_InputBindComposite(mapName, actionName, partsJson, deviceId);
    }

    public static bool SetBindingInteraction(string mapName, string actionName,
        InputInteraction interaction, float holdSeconds = 0.5f)
    {
        return NativeCalls.Native_InputSetBindingInteraction(mapName, actionName, (int)interaction, holdSeconds);
    }

    public static bool SetBindingTapParams(string mapName, string actionName, int bindingIndex, int tapCount, float tapWindowSeconds)
    {
        return NativeCalls.Native_InputSetBindingTapParams(mapName, actionName, bindingIndex, tapCount, tapWindowSeconds);
    }

    public static bool SetBindingRepeatParams(string mapName, string actionName, int bindingIndex, float repeatDelay, float repeatRate)
    {
        return NativeCalls.Native_InputSetBindingRepeatParams(mapName, actionName, bindingIndex, repeatDelay, repeatRate);
    }

    public static bool SetBindingProcessor(string mapName, string actionName, int bindingIndex, InputProcessorType type, float a = 0.0f, float b = 1.0f)
    {
        return NativeCalls.Native_InputSetBindingProcessor(mapName, actionName, bindingIndex, (int)type, a, b);
    }

    public static bool LoadActionAsset(string path) => NativeCalls.Native_InputLoadActionAsset(path);

    public static bool LoadConfigOverrides(string projectPath, string userPath)
    {
        return NativeCalls.Native_InputLoadConfigOverrides(projectPath, userPath);
    }

    public static bool SaveUserConfigOverrides(string userPath)
    {
        return NativeCalls.Native_InputSaveUserConfigOverrides(userPath);
    }

    public static uint FindActionId(string mapName, string actionName) => NativeCalls.Native_InputFindActionId(mapName, actionName);

    public static uint FindActionId(string qualifiedName) => NativeCalls.Native_InputFindActionId(qualifiedName);

    public static bool IsActionDown(string actionName)
    {
        return NativeCalls.Native_InputIsActionDown(actionName);
    }

    public static bool WasActionPressed(string actionName)
    {
        return NativeCalls.Native_InputWasActionPressed(actionName);
    }

    public static bool WasActionReleased(string actionName)
    {
        return NativeCalls.Native_InputWasActionReleased(actionName);
    }

    public static float GetActionAxis(string actionName)
    {
        return NativeCalls.Native_InputGetActionAxis(actionName);
    }

    public static Vector2f GetActionVector2(string actionName)
    {
        return NativeCalls.Native_InputGetActionVector2(actionName);
    }

    public static bool IsActionDown(uint actionId) => NativeCalls.Native_InputIsActionDown(actionId);

    public static bool WasActionPressed(uint actionId) => NativeCalls.Native_InputWasActionPressed(actionId);

    public static bool WasActionReleased(uint actionId) => NativeCalls.Native_InputWasActionReleased(actionId);

    public static float GetActionAxis(uint actionId) => NativeCalls.Native_InputGetActionAxis(actionId);

    public static Vector2f GetActionVector2(uint actionId) => NativeCalls.Native_InputGetActionVector2(actionId);

    public static Vector2f GetMousePosition()
    {
        return NativeCalls.Native_InputGetMousePosition();
    }

    public static Vector2f GetMouseDelta()
    {
        return NativeCalls.Native_InputGetMouseDelta();
    }

    public static Vector2f GetMouseWheel()
    {
        return NativeCalls.Native_InputGetMouseWheel();
    }

    public static bool IsGamepadConnected(uint deviceId) => NativeCalls.Native_InputIsGamepadConnected(deviceId);

    public static bool IsGamepadButtonDown(uint deviceId, GamepadButtonCode button) => NativeCalls.Native_InputIsGamepadButtonDown(deviceId, (int)button);

    public static bool IsGamepadButtonPressed(uint deviceId, GamepadButtonCode button) => NativeCalls.Native_InputIsGamepadButtonPressed(deviceId, (int)button);

    public static bool IsGamepadButtonReleased(uint deviceId, GamepadButtonCode button) => NativeCalls.Native_InputIsGamepadButtonReleased(deviceId, (int)button);

    public static float GetGamepadAxis(uint deviceId, GamepadAxisCode axis) => NativeCalls.Native_InputGetGamepadAxis(deviceId, (int)axis);

    public static ulong Subscribe(string actionName, InputActionPhase phase, Action<InputActionEvent> callback)
    {
        ulong token = NativeCalls.Native_InputSubscribe(actionName, (int)phase);
        uint actionId = FindActionId(actionName);
        InputEventHub.Register(token, actionId, phase, callback);
        return token;
    }

    public static ulong Subscribe(uint actionId, InputActionPhase phase, Action<InputActionEvent> callback)
    {
        ulong token = NativeCalls.Native_InputSubscribe(actionId, (int)phase);
        InputEventHub.Register(token, actionId, phase, callback);
        return token;
    }

    public static void Unsubscribe(ulong subscriptionId)
    {
        NativeCalls.Native_InputUnsubscribe(subscriptionId);
        InputEventHub.Unsubscribe(subscriptionId);
    }

    public static bool SetBindingOverride(string mapName, string actionName, int bindingIndex, string bindingJson)
    {
        return NativeCalls.Native_InputSetBindingOverride(mapName, actionName, bindingIndex, bindingJson);
    }

    public static bool ClearBindingOverride(string mapName, string actionName, int bindingIndex)
    {
        return NativeCalls.Native_InputClearBindingOverride(mapName, actionName, bindingIndex);
    }

    public static bool BeginRebindSession(string mapName, string actionName, int bindingIndex)
    {
        return NativeCalls.Native_InputBeginRebindSession(mapName, actionName, bindingIndex);
    }

    public static void CancelRebindSession() => NativeCalls.Native_InputCancelRebindSession();

    public static bool IsRebindSessionActive() => NativeCalls.Native_InputIsRebindSessionActive();
}
